File: 460-lfu-cache/460-lfu-cache.java
class LFUCache {
    HashMap<Integer,DList> freqMap;
    HashMap<Integer,Node> nodeMap;
    public int cap;
    public int size;
    public int minFreq;
    public LFUCache(int capacity) {
        freqMap = new HashMap<>();
        nodeMap = new HashMap<>();
        cap = capacity;
        size = 0;
        
        freqMap.put(1, new DList());
        minFreq = 1;
    }
    
    public int get(int key) {
        if(nodeMap.containsKey(key)){
            Node node = nodeMap.get(key);
            UpdateFreqMap(node);
            return node.val;
        }
        else
            return -1;
    }
    
    public void put(int key, int value) {
        if(cap == 0)
            return;
        
        if(nodeMap.containsKey(key)){
            Node node = nodeMap.get(key);
            node.val = value;
            UpdateFreqMap(node);
            nodeMap.put(key, node);
        }
        else{
            if(size == cap){
                DList oldList = freqMap.get(minFreq);
                Node oldNode = oldList.RemoveLast();
                nodeMap.remove(oldNode.key);
                
                freqMap.put(minFreq, oldList);
                
                size--;
            }
            
            Node node = new Node();
            node.key = key;
            node.val = value;
            
            DList newList = freqMap.get(node.freq);
            newList.Add(node);
            freqMap.put(node.freq, newList);
            nodeMap.put(key, node);
            minFreq = node.freq;
            size++;
        }
    }
    
    private void UpdateFreqMap(Node node){
        DList oldList = freqMap.get(node.freq);
        oldList.Remove(node);
        nodeMap.remove(node.key);
        
        if(oldList.size == 0 && node.freq == minFreq){
            minFreq++;
        }
        
        node.freq++;
        
        DList newList = freqMap.get(node.freq);
        if(newList == null){
            newList = new DList();
        }
        
        newList.Add(node);
        freqMap.put(node.freq, newList);
        nodeMap.put(node.key, node);
    }
}

class DList{
    public int size;
    public Node head;
    public Node tail;
    
    public DList(){
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
    }
    
    public void Add(Node node){
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;
        head.next = node;
        
        size++;
    }
    
    public void Remove(Node node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
        
        size--;
    }
    
    public Node RemoveLast(){
        Node node = tail.prev;
        Remove(tail.prev);
        return node;
    }
}

class Node{
    public int freq = 1;
    public int key;
    public int val;
    public Node next;
    public Node prev;
    
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 460-lfu-cache/460-lfu-cache.cs
public class LFUCache {
    Dictionary<int, DList> freqMap;
    Dictionary<int, Node> nodeMap;
    public int minFreq;
    public int cap;
    public int size;
    public LFUCache(int capacity) {
        this.minFreq = 1;
        this.freqMap = new Dictionary<int, DList>();
        this.freqMap.Add(1, new DList());
        
        this.nodeMap = new Dictionary<int,Node>();
        this.cap = capacity;
        this.size = 0;
    }
    
    public int Get(int key) {
        if(nodeMap.ContainsKey(key)){
            Node node = this.nodeMap[key];
            this.UpdateNodeInFreqMap(node);
            return node.val;
        }
        
        return -1;
    }
    
    public void Put(int key, int value) {
        if(cap == 0)
            return;
        
        if(this.nodeMap.ContainsKey(key)){
            Node node = this.nodeMap[key];
            node.val = value;
            this.UpdateNodeInFreqMap(node);
        }
        else{
            if(this.size == this.cap){
                DList oldList = this.freqMap[this.minFreq];
                Node lastNode = oldList.GetLastNode();
                if(lastNode != null){
                    this.nodeMap.Remove(lastNode.key);
                }
                
                oldList.RemoveLast();
                this.freqMap[this.minFreq] = oldList;
                this.size--;
                
                if(oldList.Length == 0){
                    this.minFreq++;
                }
            }
            
            Node node = new Node(key, value);
            
            DList existingList = this.freqMap[node.freq];
            existingList.AddNode(node);
            this.freqMap[node.freq] = existingList;
            this.nodeMap.Add(key, node);
            this.minFreq = node.freq;
            this.size++;
        }
    }
    
    private void UpdateNodeInFreqMap(Node node){
        DList list = freqMap[node.freq];
        list.RemoveNode(node);
        
        if(list.Length == 0 && node.freq == this.minFreq){
            this.minFreq++;
        }
        
        node.freq++;
        
        if(!freqMap.ContainsKey(node.freq)){
            DList newList = new DList();
            newList.AddNode(node);
            this.freqMap.Add(node.freq, newList);
        }
        else{
            DList existingList = this.freqMap[node.freq];
            existingList.AddNode(node);
        }
        
        this.nodeMap[node.key] = node;
    }
}


public class DList{
    public int Length;
    public Node head;
    public Node tail;
    
    public DList(){
        this.head = new Node();
        this.tail = new Node();
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }
    
    public void AddNode(Node node){
        //adding to the front
        node.next = this.head.next;
        node.prev = this.head;
        this.head.next.prev = node;
        this.head.next = node;
        
        this.Length++;
    }
    
    public void RemoveNode(Node node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
        
        this.Length--;
    }
    
    public void RemoveLast(){
        if(this.Length > 0)
            RemoveNode(this.tail.prev);
    }
    
    public Node GetLastNode(){
        if(this.Length > 0)
            return this.tail.prev;
        
        return null;
    }
    
}


public class Node{
    public int key;
    public int val;
    public int freq;
    public Node next;
    public Node prev;
    
    public Node(int _key = 0, int _val = 0, Node _next = null, Node _prev = null){
        this.key = _key;
        this.val = _val;
        this.next = _next;
        this.prev = _prev;
        this.freq = 1;
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */